### PR TITLE
Add unit tests for various edge cases

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR adds some extra unit tests for various edge cases. It also fixes (one of) the edge cases that has validation which is currently unreachable. It also changes a few stray `sys.exit()` calls to just raise errors, since I don't think there's a special reason to have those there at the moment.

NOTE: Some of these tests are testing errors that are thrown by HF libraries for things we don't explicitly validate ourselves to ensure the behavior is stable with version bumps.

### Related issue number
This is a follow-up on: https://github.com/foundation-model-stack/fms-hf-tuning/issues/74

We made the decision to split the PR into two parts, to get the tests covering the base functionality it earlier.

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: tests/test_sft_trainer.py [hunk 1]
```diff
@@ -323,3 +327,191 @@ def _validate_adapter_config(adapter_config, peft_type, base_kwargs):
...
+        model_args, data_args, training_args, tune_config = causal_lm_train_kwargs(
+            TRAIN_KWARGS
+        )
+        with pytest.raises(ValueError):
+            sft_trainer.train(model_args, data_args, training_args, tune_config)
+
+
+### Tests for model dtype edge cases
+@pytest.mark.skipif(
+    not (torch.cuda.is_available() and torch.cuda.is_bf16_supported()),
```
Comment (alex-jw-brooks) [inline tests/test_sft_trainer.py:488]: `cuda` check is here needed because the bf16 check throws if no Nvidia drivers are available

File: tests/test_sft_trainer.py [hunk 2]
```diff
@@ -323,3 +327,191 @@ def _validate_adapter_config(adapter_config, peft_type, base_kwargs):
...
+    """Ensure that we get a FileNotFoundError if the data is missing completely."""
+    TRAIN_KWARGS = {
+        **BASE_PEFT_KWARGS,
+        **{"training_data_path": "/foo/bar/foobar", "output_dir": "foo/bar/baz"},
+    }
+    model_args, data_args, training_args, tune_config = causal_lm_train_kwargs(
+        TRAIN_KWARGS
+    )
+    with pytest.raises(FileNotFoundError):
+        sft_trainer.train(model_args, data_args, training_args, tune_config)
```
Comment (anhuong) [inline tests/test_sft_trainer.py:418]: this test already exists -- https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tests/test_sft_trainer.py

File: tests/test_sft_trainer.py [hunk 3]
```diff
@@ -323,3 +327,191 @@ def _validate_adapter_config(adapter_config, peft_type, base_kwargs):
...
+            sft_trainer.train(model_args, data_args, training_args, tune_config)
+
+
+### Tests for model dtype edge cases
+@pytest.mark.skipif(
+    not (torch.cuda.is_available() and torch.cuda.is_bf16_supported()),
+    reason="Only runs if bf16 is unsupported",
+)
+def test_bf16_still_tunes_if_unsupported():
+    """Ensure that even if bf16 is not supported, tuning still works without problems."""
```
Comment (anhuong) [inline tests/test_sft_trainer.py:492]: interesting test case! can you explain why it doesn't fail and tuning still works and why this is the preferred expected behavior?

**Review by anhuong — COMMENTED** [review]
Great test cases, thanks Alex! Left a few questions...

File: tests/test_sft_trainer.py (hunk 3 above)
Comment (alex-jw-brooks) [reply tests/test_sft_trainer.py:492]: As far as I understand, in devices where bfloat16 is unsupported, there is usually fallback behavior to a supported data type, which is usually float32 since bfloat16 and float32 have the same exponent size!

File: tuning/sft_trainer.py [hunk 4]
```diff
@@ -194,6 +193,26 @@ def train(
...
+    if train_args.packing:
+        logger.info("Packing is set to True")
+        data_collator = None
+        packing = True
+    else:
+        logger.info("Packing is set to False")
+        if data_args.response_template is None:
+            # TODO: Fix this, currently unreachable due to crashing in batch encoding tokenization
+            # We should do this validation up front, then do the encoding, then handle the collator
+            raise ValueError("Response template is None, needs to be set for training")
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply tuning/sft_trainer.py:214]: Yeah, you can't encode a None type with a tokenizer since tokenizers generally expect an input of type ` Union[TextInputSequence, Tuple[InputSequence, InputSequence]]`. It would be best to do that in a separate PR to keep things atomic even though it's a simple change, since some of the validation logic is a little bit delicate

**Review by anhuong — APPROVED** [review]
Other than the linter error for a line being too long, the tests look good! Thanks Alex
```
tests/test_sft_trainer.py:422:0: C0301: Line too long (105/100) (line-too-long)
```

**Review by anhuong — APPROVED** [review]
Thanks Alex!
